Python/traceback.c

```diff
@@ -554,29 +696,53 @@ tb_displayline(PyTracebackObject* tb, PyObject *f, PyObject *filename, int linen
         if (start_col_byte_offset < 0 || end_col_byte_offset < 0) {
             goto done;
         }
+
         // Convert the utf-8 byte offset to the actual character offset so we
         // print the right number of carets.
-        Py_ssize_t start_offset = _PyPegen_byte_offset_to_character_offset(source_line, start_col_byte_offset);
-        Py_ssize_t end_offset = _PyPegen_byte_offset_to_character_offset(source_line, end_col_byte_offset);
+        Py_ssize_t start_offset = (Py_ssize_t)start_col_byte_offset;
+        Py_ssize_t end_offset = (Py_ssize_t)end_col_byte_offset;
 
-        char offset = truncation;
-        while (++offset <= start_offset) {
-            err = PyFile_WriteString(" ", f);
-            if (err < 0) {
-                goto done;
+        if (source_line) {
+            start_offset = _PyPegen_byte_offset_to_character_offset(source_line, start_col_byte_offset);
+            end_offset = _PyPegen_byte_offset_to_character_offset(source_line, end_col_byte_offset);
+        }
+
+        const char *primary, *secondary;
+        primary = secondary = "^";
+
+        int left_end_offset, right_start_offset;
+        left_end_offset = right_start_offset = Py_SAFE_DOWNCAST(end_offset, Py_ssize_t, int) - Py_SAFE_DOWNCAST(start_offset, Py_ssize_t, int);
```

Comment:
I would just have `extract_anchors_from_line` take `Py_ssize_t`s as parameters if possible so you don't have to do this.

Reply:
then we need to cast each ast offset (regular ints) at the place.

Reply:
Aah, I think in practice `int` should almost always be safe to assign to `Py_ssize_t`, going by the lack of compiler complaints and https://github.com/python/cpython/blob/4d4be477055d0f9e415407bcd193788fae04dc11/Include/pyport.h#L127-L130

Theoretically `ssize_t` nor `size_t` is not mandated to have any relation to `int`, so it might be possible for there to be a platform where it's not safe to do, feel free to go with whichever one you feel more comfortable with.